addons/crm/models/crm_lead.py

```diff
@@ -1434,11 +1503,146 @@ def _update_automated_probabilities(self):
             )
         )
 
-    # ----------------------------
+    # ---------------------------------
+    # PLS: Common parts for both mode
+    # ---------------------------------
+    def _pls_prepare_update_frequency_table(self, rebuild=False, target_state=False):
+        """
+        This method is common to Live Increment or Full Rebuild mode, as it shares the main steps.
+        This method will prepare the frequency dict needed to update the frequency table:
+            - New frequencies: frequencies that we need to add in the frequency table.
+            - Existing frequencies: frequencies that are already in the frequency table.
+        In rebuild mode, only the new frequencies are needed as existing frequencies are truncated.
+        For each team, each dict contains the frequency in won and lost for each field/value couple
+        of the target leads.
+        Target leads are :
+            - in Live increment mode : given ongoing leads (self)
+            - in Full rebuild mode : all the closed (won and lost) leads in the DB.
+        During the frequencies update, with both new and existing frequencies, we can split frequencies to update
+        and frequencies to add. If a field/value couple already exists in the frequency table, we just update it.
+        Otherwise, we need to insert a new one.
+        """
+        # Keep eligible leads
+        pls_start_date = self._pls_get_safe_start_date()
+        if not pls_start_date:
+            return {}, {}
+
+        if rebuild:  # rebuild will treat every closed lead in DB, increment will treat current ongoing leads
+            pls_leads = self
+        else:
+            # Only treat leads created after the PLS start Date
+            pls_leads = self.filtered(
+                lambda lead: fields.Date.to_date(pls_start_date) <= fields.Date.to_date(lead.create_date))
+            if not pls_leads:
+                return {}, {}
+
+        # Extract target leads values
+        team_ids = self.env['crm.team'].with_context(active_test=False).search([]).ids + [0]  # If team_id is unset, consider it as team 0
```

Comment:
Why take all the team if you are not in rebuild mode. only the team of the current lead should be enough see below